README.md

```diff
@@ -1,7 +1,123 @@
-3. Analysis of MRL models via superclass hierarchy and oracle performance
+This repository contains the code for training and evaluating Matryoshka Representation with ResNet-50 backbone, along with visualizations. Training code is built upon [FFCV](https://github.com/libffcv/ffcv-imagenet) modified for MRL. The repository is organized as follows:
 
+1. Set up
+2. Matryoshka Linear Layer
+3. Training ResNet50 Models
+4. Evaluating Trained Models
+5. Evaluating pre-trained checkpoints
+5. Retrieval Performance
+
+
+## Set Up
+First pip install the requirements file in this directory:
+```
+pip install -r requirements.txt
+```
+We use PyTorch Distributed Data Parallel shared over 2 A100 GPUs and FFCV for the dataloader. FFCV uses 8 A100 GPUs, therefore we linearly downscale the learning rate by 4.
+
+### Preparing the dataset.
+
+Generate ImageNet FFCV dataset from standard imagenet train set with the following command (`IMAGENET_DIR` should point to a PyTorch style [ImageNet dataset](https://github.com/MadryLab/pytorch-imagenet-dataset)):
+
+```bash
+# Required environmental variables for the script:
+export IMAGENET_DIR=/path/to/pytorch/format/imagenet/directory/
+export WRITE_DIR=/your/path/here/
+
+# Serialize images with:
+# - 500px side length maximum
+# - 50% JPEG encoded, 90% raw pixel values
+# - quality=90 JPEGs
+.train_and_eval/train/write_imagenet.sh 500 0.50 90
+```
+
+#### Robustness Datasets
+
+We evaluate our trained models on 4 robustness datasets- ImageNet-V2/A/R/Sketch. Please refer to the respective repositories to download the datasets and additional documentation. 
+
+1. [ImageNetV2_pytorch](https://github.com/modestyachts/ImageNetV2_pytorch)
+2. [ImageNetA](https://github.com/hendrycks/natural-adv-examples)
+3. [ImageNetR](https://github.com/hendrycks/imagenet-r)
+4. [ImageNet-Sketch](https://github.com/HaohanWang/ImageNet-Sketch)
+
+
+## Matryoshka Linear Layer
+`MRL.py` provides MRL linear layer, and it can be instantiated very easily using the following
+```
+nesting_list=[8, 16, 32, 64, 128, 256, 512, 1024, 2048]
+fc_layer = MRL_Linear_Layer(nesting_list, num_classes=1000, efficient=efficient)
+```
+Where, `nesting_list` is the list of dimensions we want to nest on, `num_classes` is the number of output feature and efficient flag if we want to do MRL-E.
+
+## Training ResNet50 models
+First of all change the directory to `train_and_eval/train`. While training, we dump  model ckpts, training logs and run command inside a folder. 
+
+### Training Fixed Feature Baseline
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1
+
+python train_imagenet.py --config-file rn50_configs/rn50_40_epochs.yaml --model.fixed_feature=2048 --data.train_dataset=$WRITE_DIR/train_500_0.50_90.ffcv --data.val_dataset=$WRITE_DIR/val_500_uncompressed.ffcv --data.num_workers=12 --data.in_memory=1 --logging.folder=trainlogs --logging.log_level=1 --dist.world_size=2 --training.distributed=1 --lr.lr=0.425
```

Comment:
Where do logs go? ALso, what is $Write_DIR shouldn't it be $Data_dir?

Reply:
There is a separate folder for train logs where we dump logs. Can change $WRITE_DIR to $DATA_DIR but it was done to make sure we have consistency with FFCV notations.